### PR TITLE
add error handling to CI formatting check

### PR DESCRIPTION
**Pull request by efx**
Start addressing #1129.
I added error handling flag `-e` and shellcheck lints.
I'll push up fixes for the associated failures.

### CODE REVIEW

File: bin/format_exercises [hunk 1]
```diff
@@ -1,17 +1,17 @@
 #!/bin/sh
 # Format existing exercises using rustfmt
+set -e
 
 RUST_TRACK_REPO_PATH=$(cd "$(dirname "$0")/.." && pwd)
 
-EXERCISES_PATH="${RUST_TRACK_REPO_PATH}/exercises/"
+EXERCISES_PATH="${RUST_TRACK_REPO_PATH}/exercises/practice"
```
Comment (coriolinus) [inline bin/format_exercises:7]:
```suggestion
EXERCISES_PATH="${RUST_TRACK_REPO_PATH}/exercises/*"
```

I don't see any reason to exclude concept exercises from this check.

File: bin/format_exercises [hunk 2]
```diff
@@ -1,17 +1,17 @@
 #!/bin/sh
 # Format existing exercises using rustfmt
+set -e
```
Comment (coriolinus) [inline bin/format_exercises:3]: 👍 If we ever do script lints, I'd want to see this line pretty much ubiquitously in our scripts. Saves a lot of errors.

File: bin/format_exercises [hunk 3]
```diff
@@ -1,17 +1,17 @@
...
-  exercise_name=$(basename $exercise_dir)
+for exercise_dir in "${EXERCISES_PATH}"/*; do
   (
     cd "$exercise_dir"
     cargo fmt
-    if [ -f example.rs ]; then
-      rustfmt example.rs
+    if [ -f .meta/example.rs ]; then
+      rustfmt .meta/example.rs
     fi
```
Comment (coriolinus) [inline bin/format_exercises:15]: We need to duplicate this check for `.meta/exemplar.rs`

File: exercises/practice/accumulate/src/lib.rs [hunk 4]
```diff
@@ -1,4 +1,7 @@
 /// What should the type of _function be?
-pub fn map(input: Vec<i32>, _function: ???) -> Vec<i32> {
+pub fn map<F, T, U>(input: Vec<i32>, _function: F) -> Vec<i32>
+where
+    F: FnMut(T) -> U,
+{
```
Comment (coriolinus) [inline exercises/practice/accumulate/src/lib.rs:5]: Strongly negative on this change. If rustfmt is complaining about it, we can add a meta-skip check for rustfmt for this exercise; see `.meta/ALLOWED_TO_NOT_COMPILE` for the compile checks for an example of how that's handled.

Comment (efx) [reply exercises/practice/accumulate/src/lib.rs]: Got it! I naively copied this but see that it should be invalid Rust code. I reverted this change and ensured `format_exercises` skips formatting on exercises with `.meta/ALLOWED_TO_NOT_COMPILE`

File: bin/format_exercises [hunk 5]
```diff
@@ -1,17 +1,36 @@
...
+# directory and format Rust files
+format_exercises() {
+    EXERCISES_PATH="${RUST_TRACK_REPO_PATH}/exercises/$1"
+    source_file_name="$2"
+    for exercise_dir in "${EXERCISES_PATH}"/*; do
+        (
+        cd "$exercise_dir"
+	    allowed_file="$exercise_dir"/.meta/ALLOWED_TO_NOT_COMPILE
+	    if [ -f "$allowed_file" ]; then
+	        exercise=$(basename "$exercise_dir")
```
Comment (coriolinus) [inline bin/format_exercises:17]: Indentation got weird here.

**Review by coriolinus — APPROVED** [review]
Aside from some formatting weirdness in the shell script--a little ironic, given the topic of the PR--LGTM.

File: bin/format_exercises (hunk 5 above)
Comment (efx) [reply bin/format_exercises]: Nice catch! My vim indentation messes this up despite editorconfig. Inspires me to enforce soon in CI!

**Pull request merged**